fastn-core/src/auth/emailpassword.rs

```diff
@@ -0,0 +1,371 @@
+use argon2::{Argon2, PasswordHasher, PasswordVerifier};
+
+#[derive(Debug)]
+pub struct Token {
+    pub session_id: uuid::Uuid,
+    pub token: uuid::Uuid,
+    pub provider: String,
+}
+
+impl Token {
+    fn from_row(row: &tokio_postgres::Row) -> Self {
+        Token {
+            session_id: row.get("session_id"),
+            token: row.get("token"),
+            provider: row.get("provider"),
+        }
+    }
+}
+
+// TODO: handle errors
+pub async fn create_user(
+    req: &fastn_core::http::Request,
+) -> fastn_core::Result<fastn_core::http::Response> {
+    if req.method() != "POST" {
+        return Ok(fastn_core::not_found!("invalid route"));
+    }
+
+    #[derive(serde::Deserialize, serde::Serialize, Debug)]
+    struct UserPayload {
+        username: String,
+        email: String,
+        name: String,
+        password: String,
+    }
+
+    let user: UserPayload = serde_json::from_slice(req.body())?;
+
+    tracing::info!("user payload: {:?}", &user);
+
+    let client = db::get_client().await?;
+
+    let res = client
+        .query(
+            "select count(id) from fastn_user where username = $1 or email = $2",
+            &[&user.username, &user.email],
+        )
+        .await
+        .unwrap();
+
+    let user_count: i64 = res[0].get(0);
+
+    if user_count > 0 {
+        // TODO: use correct HTTP status code
+        return fastn_core::http::api_error("user already exists");
+    }
+
+    // encrypt password
+    let salt =
+        argon2::password_hash::SaltString::generate(&mut argon2::password_hash::rand_core::OsRng);
+
+    let argon2 = argon2::Argon2::default();
+
+    let hashed_password = argon2
+        .hash_password(user.password.as_bytes(), &salt)
+        .unwrap()
+        .to_string();
+
+    let user_id = uuid::Uuid::new_v4();
+
+    // TODO: email verification
+    // TODO: password can be null for oauth users. if it's null they should not login using
+    // emailpassword flow
+    let user = upsert_user(
+        &user_id,
+        user.email.as_str(),
+        user.username.as_str(),
+        user.name.as_str(),
+        hashed_password.as_str(),
+    )
+    .await?;
+
+    tracing::info!("fastn_user created. user: {:?}", &user);
+
+    fastn_core::http::api_ok(user)
```

Comment:
On success we return redirect to the right place.

Reply:
Are you sure? This `POST` requests will be done by clients (`fetch` for example) and they'll have to [handle the redirect anyways](https://stackoverflow.com/a/39739894), shouldn't we just send the json response and let them handle the redirection?

Reply:
But you are not sending JSON with redirection info, you are sending user data serialised as JSON. What will client do with non redirect json?